internal/checker/checker.go

```diff
@@ -21174,14 +21176,64 @@ func (c *Checker) instantiateTypeWithAlias(t *Type, m *TypeMapper, alias *TypeAl
 		c.error(c.currentNode, diagnostics.Type_instantiation_is_excessively_deep_and_possibly_infinite)
 		return c.errorType
 	}
+	index := c.findActiveMapper(m)
+	if index == -1 {
+		c.pushActiveMapper(m)
+	}
+	var b KeyBuilder
+	b.WriteType(t)
+	b.WriteAlias(alias)
+	key := b.String()
+	cache := c.activeTypeMappersCaches[core.IfElse(index != -1, index, len(c.activeTypeMappersCaches)-1)]
+	if cachedType, ok := cache[key]; ok {
+		return cachedType
+	}
 	c.TotalInstantiationCount++
 	c.instantiationCount++
 	c.instantiationDepth++
 	result := c.instantiateTypeWorker(t, m, alias)
+	if index == -1 {
+		c.popActiveMapper()
+	} else {
+		cache[key] = result
+	}
 	c.instantiationDepth--
 	return result
 }
 
+var activeMapperCache = sync.Pool{
+	New: func() any {
+		return make(map[string]*Type, 1)
+	},
+}
+
+func (c *Checker) pushActiveMapper(mapper *TypeMapper) {
+	c.activeMappers = append(c.activeMappers, mapper)
+
+	c.activeTypeMappersCaches = append(c.activeTypeMappersCaches, activeMapperCache.Get().(map[string]*Type))
+}
+
+func (c *Checker) popActiveMapper() {
+	c.activeMappers[len(c.activeMappers)-1] = nil
```

Comment:
FWIW this nil-ing out is _not_ done in Strada, but should be, because the stacks in Strada are always held onto. I'm going to send a PR to fix that in Strada.

Reply:
if I understand the concern correctly, this stack was modeled after the one that deals with contextual types - `popContextualType` doesn't clear up stale entries either. My assumption was the claim is that those slots will (likely) eventually be replaced by other entries anyway so it's not worth clearing the stack entry itself. But with those mappers the situation might be slightly more delicate as the mapper cache might statisictally hold onto more objects than a contextual type

Reply:
Ah, you're right, and those are wrong in the Go code too.

Reply:
https://github.com/microsoft/TypeScript/pull/62016